### PR TITLE
Migrate from using a GH personal access token to a GH app

### PR DESCRIPTION
**Pull request by jljaco**
Issue #, if available:  https://github.com/aws-controllers-k8s/community/issues/1565

Description of changes:

Update Prow deployments to use a GitHub app instead of a GitHub personal access token.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: flux/prow-config.yaml
Comment (RedbackThomson) [inline flux/prow-config.yaml:26]: Not sure what this is, but this value should be loaded in from a `Secret` in the same way we load the PAT today.

File: infra/README.md
Comment (RedbackThomson) [inline infra/README.md:28]: nit: this can be a `cat` of the PEM file?

File: infra/lib/charts/prow-secrets.ts
Comment (RedbackThomson) [inline infra/lib/charts/prow-secrets.ts:29]: nit: Follow same casing rules as earlier (`appId`)

File: prow/config/Chart.yaml
Comment (RedbackThomson) [inline prow/config/Chart.yaml:5]: 👍🏼

File: prow/config/templates/crier-Deployment.yaml
Comment (RedbackThomson) [inline prow/config/templates/crier-Deployment.yaml:58]: nit: I'd prefer if we could stick to the same casing standards here (make it `appId`)

File: prow/config/templates/github-token-Secret.yaml
Comment (RedbackThomson) [inline prow/config/templates/github-token-Secret.yaml:24]: See casing notes as above

File: prow/config/templates/horologium-Role.yaml
Comment (RedbackThomson) [inline prow/config/templates/horologium-Role.yaml:27]: Weird whitespacing here?

**Review by RedbackThomson — CHANGES_REQUESTED** [review]

File: prow/config/templates/config-ConfigMap.yaml
Comment (A-Hilaly) [inline prow/config/templates/config-ConfigMap.yaml:109]: I guess there is a merge conflict as this was addressed in https://github.com/aws-controllers-k8s/test-infra/pull/262

File: infra/lib/ci-cluster.ts
Comment (A-Hilaly) [inline infra/lib/ci-cluster.ts:111]: Nice! Looks like the version we were using was released 18months ago.. [1f0ff4b46a99155089b2bd19e9c0c2acad2069cc](https://github.com/bitnami/charts/commit/1f0ff4b46a99155089b2bd19e9c0c2acad2069cc)

File: prow/config/templates/horologium-Role.yaml
Comment (A-Hilaly) [reply prow/config/templates/horologium-Role.yaml:27]: Just Github rendering 🤷

File: prow/config/templates/prow-controller-manager-Deployment.yaml
Comment (A-Hilaly) [inline prow/config/templates/prow-controller-manager-Deployment.yaml:49]: Interesting! deck is no longer needed?

File: prow/config/templates/plugins-ConfigMap.yaml
Comment (A-Hilaly) [inline prow/config/templates/plugins-ConfigMap.yaml:49]: To keep to scope of this PR simple and precise. Can we make this in a separate PR? We'll need other plugins after #263 is merged too.

File: prow/config/values.yaml
Comment (A-Hilaly) [inline prow/config/values.yaml:22]: Same comment here, maybe we can raise a PR Just for bumping images version. Just in case something breaks it's gonna be easier to debug (and worst case revert)

**Review by A-Hilaly — COMMENTED** [review]
Good stuff @jljaco ! Thank you!

Left few comments and questions below

File: infra/README.md
Comment (jljaco) [reply infra/README.md:28]: Yes it can.  I added an example at he end of the README.md illustrating how one can use command substitution to do that.

File: infra/lib/charts/prow-secrets.ts
Comment (jljaco) [reply infra/lib/charts/prow-secrets.ts:29]: Won't do -- the [deployment Getting Started guide](https://github.com/kubernetes-sigs/prow/blob/main/site/content/en/docs/Legacy%20Snapshot/prow/getting_started_deploy.md#create-the-github-secrets) uses the `appid` spelling, so I think it's best to stay consistent.

File: prow/config/templates/config-ConfigMap.yaml
Comment (jljaco) [reply prow/config/templates/config-ConfigMap.yaml:109]: You found that and fixed it before I got to it!  Nicely done

File: prow/config/templates/crier-Deployment.yaml
Comment (jljaco) [reply prow/config/templates/crier-Deployment.yaml:58]: Won't do (see above)

File: prow/config/templates/github-token-Secret.yaml
Comment (jljaco) [reply prow/config/templates/github-token-Secret.yaml:24]: Won't do

File: prow/config/templates/horologium-Role.yaml
Comment (jljaco) [reply prow/config/templates/horologium-Role.yaml:27]: Yep, that's just GitHub being weird

File: prow/config/templates/prow-controller-manager-Deployment.yaml
Comment (jljaco) [reply prow/config/templates/prow-controller-manager-Deployment.yaml:49]: Apparently not!  peep https://raw.githubusercontent.com/kubernetes/test-infra/master/config/prow/cluster/starter/starter-s3.yaml and you will see there is no longer any mention of `deck-url`

File: infra/lib/charts/prow-secrets.ts
Comment (RedbackThomson) [reply infra/lib/charts/prow-secrets.ts:29]: I am fine with it being `appid` in the Secret, to stay consistent with their secret, but we should at least stick to our casing for the `values.yaml` file.

File: prow/config/templates/plugins-ConfigMap.yaml
Comment (jljaco) [reply prow/config/templates/plugins-ConfigMap.yaml]: Done

File: prow/config/values.yaml
Comment (jljaco) [reply prow/config/values.yaml:22]: If I don't bump the version in this PR, then everything will be broken until the other PR is merged.  Shouldn't we aim to make PRs idempotent?

File: flux/prow-config.yaml
Comment (jljaco) [reply flux/prow-config.yaml:26]: Good catch, I had been using it for setting the app ID manually during testing.

Removed.

File: infra/lib/charts/prow-secrets.ts
Comment (jljaco) [reply infra/lib/charts/prow-secrets.ts:29]: OK -- I think it's actually more confusing this way, but I'll defer to your recommendation here.

Done

File: prow/config/templates/crier-Deployment.yaml
Comment (jljaco) [reply prow/config/templates/crier-Deployment.yaml]: Done

File: prow/config/templates/github-token-Secret.yaml
Comment (jljaco) [reply prow/config/templates/github-token-Secret.yaml]: Done

File: prow/config/values.yaml
Comment (A-Hilaly) [reply prow/config/values.yaml:22]: Ohhh gotcha! Well in the case let's keep them this way

File: prow/config/templates/crier-Deployment.yaml
Comment (jljaco) [reply prow/config/templates/crier-Deployment.yaml:58]: https://github.com/jljaco/test-infra/blob/f885e4e03715687d184a25fa096a4b8d19d24c29/prow/config/templates/github-token-Secret.yaml#L15 will skip secret creation under normal circumstances

https://github.com/jljaco/test-infra/blob/f885e4e03715687d184a25fa096a4b8d19d24c29/infra/lib/charts/prow-secrets.ts#L26 is responsible for creating the secret as part of the CDK deployment

I am currently in the process of doing a 100% from-scratch run through which should include that step; will let you know how that goes.

File: prow/jobs/jinja/postsubmits/codegen_release.jinja2
Comment (RedbackThomson) [inline prow/jobs/jinja/postsubmits/codegen_release.jinja2:28]: I'm not sure if our scripts, which use the Github PAT to authenticate git pull/push, will be able to support using a cert to run their jobs. This needs extra validation

**Review by RedbackThomson — CHANGES_REQUESTED** [review]

File: prow/config/templates/crier-Deployment.yaml
Comment (jljaco) [reply prow/config/templates/crier-Deployment.yaml]: Done

File: prow/jobs/jinja/postsubmits/codegen_release.jinja2
Comment (jljaco) [reply prow/jobs/jinja/postsubmits/codegen_release.jinja2:28]: Your'e right, [we need to migrate](https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps#authenticating-as-a-github-app) from including the Personal Access Token as part of request URLs ([example](https://github.com/aws-controllers-k8s/test-infra/blob/76e2cacd611c5054baa1636363c7b5e0014efaa4/cd/auto-generate/auto-generate-controllers.sh#L253)) to generating JSON web tokens and include them via an `Authorization:` HTTP header.

From a cursory examination of `test-infra/cd/**` it appears that all of the scripts there are in bash.  Are there any non-bash script-like code that you're aware of that uses the PATs?    @A-Hilaly would appreciate your input here as well.

Comment (jljaco) [reply prow/jobs/jinja/postsubmits/codegen_release.jinja2:28]: FWIW these are the places I've found so far that appear to use a PAT:
https://github.com/aws-controllers-k8s/community/blob/e001fa05183b32d8d2bcbcfc6c52c9a9864b0eb2/docs/scripts/gen_services.py#L100
https://github.com/aws-controllers-k8s/test-infra/blob/76e2cacd611c5054baa1636363c7b5e0014efaa4/cd/auto-generate/auto-generate-controllers.sh#L253
https://github.com/aws-controllers-k8s/test-infra/blob/2bc4f6fd193e2c6b2d91cc25f93acb976d3e8f37/cd/auto-generate/controller-release-tag.sh#L115
https://github.com/aws-controllers-k8s/test-infra/blob/1447a1d39a79afa23f48e2a5d8b30ecd248ee05b/cd/auto-update/project-static-files.sh#L160
https://github.com/aws-controllers-k8s/test-infra/blob/1bdb5f3c260bb6dcbce92cc6c5bb7c28fc22f30a/cd/olm/olm-bundle-pr.sh#L179
https://github.com/aws-controllers-k8s/test-infra/blob/5d4d92a8c84daf77fb2bad4b03253dd1c91c1f66/prow/jobs/images/build-docs.sh#L43
https://github.com/aws-controllers-k8s/test-infra/blob/81a36ff345e15b0920cc4cb56e0308df6c96a38e/tools/cmd/ack-discover#L56

Comment (A-Hilaly) [reply prow/jobs/jinja/postsubmits/codegen_release.jinja2:28]: >Are there any non-bash script-like code that you're aware of that uses the PATs?

Not that i'm aware of in `test-infra`, however ack-dev uses PATs in go code to interact with Github.

>Your'e right, [we need to migrate](https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps#authenticating-as-a-github-app) from including the Personal Access Token as part of request URLs ([example](https://github.com/aws-controllers-k8s/test-infra/blob/76e2cacd611c5054baa1636363c7b5e0014efaa4/cd/auto-generate/auto-generate-controllers.sh#L253)) to generating JSON web tokens and include them via an Authorization: HTTP header.

@jljaco for scripts that needs to git pull/push, are you thinking of using tokens with some like: `git push https://<GITHUB_ACCESS_TOKEN>@github.com/<GITHUB_USERNAME>/<REPOSITORY_NAME>.git
`

Comment (jljaco) [reply prow/jobs/jinja/postsubmits/codegen_release.jinja2:28]: @A-Hilaly I am thinking of [using an installation access token](https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps#authenticating-as-an-installation).  Do you think that would work?

Comment (jljaco) [reply prow/jobs/jinja/postsubmits/codegen_release.jinja2:28]: Another option is to use the [`content` section of the GitHub API](https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#create-or-update-file-contents) to perform all of the equivalent actions.    There would be significantly more effort required for that.

File: infra/README.md
Comment (azpaulp) [inline infra/README.md:16]: This link 404s for me, I see a file in the repo named getting-started-deploy.md (dash vs underscore)

Comment (jljaco) [reply infra/README.md]: Fixed link, thanks!

File: prow/jobs/jinja/postsubmits/codegen_release.jinja2
Comment (jljaco) [reply prow/jobs/jinja/postsubmits/codegen_release.jinja2:28]: We (@RedbackThomson, @azpaulp, and I ) discussed this offline, and came to the conclusion that the best way forward here is to leave the PATs in for the sake of the scripts, while still converting Prow to using the GitHub App model.

File: infra/lib/charts/app-of-apps.yaml
Comment (jaypipes) [inline infra/lib/charts/app-of-apps.yaml:1]: Why was this and the three Typescript files in `infra/lib/charts/` deleted?

Comment (RedbackThomson) [reply infra/lib/charts/app-of-apps.yaml:1]: We actually weren't using 2 of those 3 files. This app-of-apps was rolled into a hardcoded struct inside the Typescript

**Comment by jljaco** [comment]
These changes have now been tested to a moderate degree; I've gone as far as validating that the presubmit hooks for a specific controller work as expected (i.e., tests were automatically launched for https://github.com/aws-controllers-k8s/s3-controller/pull/93; unit tests, release test, and e2e tests passed).

File: cd/lib/jwt.sh
Comment (RedbackThomson) [inline cd/lib/jwt.sh:27]: Add `exit 1`?

File: cd/lib/jwt.sh
Comment (RedbackThomson) [inline cd/lib/jwt.sh:46]: Could we use `local` instead of `declare`? They're basically the same, except we use `local` in our other scripts in `test-infra`

File: cd/lib/jwt.sh
Comment (RedbackThomson) [inline cd/lib/jwt.sh:53]: Could we simplify this to:
```suggestion
echo "$input" | jq -c .
```

File: infra/lib/charts/prow-secrets.ts
Comment (RedbackThomson) [inline infra/lib/charts/prow-secrets.ts:12]: Might be worth it to rename this whole class to `ProwGithubSecretsChart` and remove the word `github` from appearing 100x

**Review by RedbackThomson — APPROVED** [review]
I have some nits about naming and code style, but functionally I think we are ready to rock

File: cd/lib/jwt.sh
Comment (jljaco) [reply cd/lib/jwt.sh:27]: `usage()` already has an `exit 1` as the last statement

File: cd/lib/jwt.sh
Comment (A-Hilaly) [inline cd/lib/jwt.sh:4]: Any idea if it's ok to use code from a blog post? Is there any license attached to it? maybe we could reach out to the person and ask them?

File: infra/README.md
Comment (A-Hilaly) [inline infra/README.md:33]: nit: `<test-infra root>` is always test-infra. I wouldn't be surprised if using a different name breaks some scripts/pipelines

File: infra/bin/test-ci.ts
Comment (A-Hilaly) [inline infra/bin/test-ci.ts:27]: nit: to keep names consistent shall we rename the context names to `github_pat`, `github_app_id` etc... ?

File: infra/lib/charts/prow-secrets.ts
Comment (jljaco) [reply infra/lib/charts/prow-secrets.ts]: Done

File: cd/lib/jwt.sh
Comment (jljaco) [reply cd/lib/jwt.sh]: Done

File: cd/lib/jwt.sh
Comment (jljaco) [reply cd/lib/jwt.sh]: Done

File: prow/jobs/jinja/periodics/label_sync.jinja2
Comment (A-Hilaly) [inline prow/jobs/jinja/periodics/label_sync.jinja2:19]: Note to self, this flag will soon be deprecated.
@jljaco i see that the label_sync command has new flags (hourly tokens and allowed burst) are these possible to use with this feature? https://github.com/kubernetes/test-infra/blob/master/label_sync/main.go#L152-L153

**Review by A-Hilaly — APPROVED** [review]
Great work on this @jljaco ! I only left nit comments below.
Letting the lgtm-ing for @RedbackThomson

**Review by RedbackThomson — APPROVED** [review]
All good on my end!

File: infra/bin/test-ci.ts
Comment (jljaco) [reply infra/bin/test-ci.ts:27]: @RedbackThomson had a similar comment here: https://github.com/aws-controllers-k8s/test-infra/pull/264#discussion_r1085887624  which I addressed by putting `GitHub` in the name of the class and removing all of the repetitive `github`s in variable names

File: prow/jobs/jinja/periodics/label_sync.jinja2
Comment (jljaco) [reply prow/jobs/jinja/periodics/label_sync.jinja2:19]: I don't know about hourly tokens limits and allowed burst.  Could be something to investigate, but isn't that outside of the scope of this PR?

This flag is now used to store the GH client secret, app ID, and client ID in a secret here: https://github.com/jljaco/test-infra/blob/a957857002f11ed2cccf66599d0877cf33fe8069/infra/lib/charts/prow-secrets.ts#L62

File: cd/lib/jwt.sh
Comment (jljaco) [reply cd/lib/jwt.sh:4]: Not sure; the original blog post (from 2016) has no license attached to the code.  

TBH this script is not even necessary for this change as it stands now (was a remnant from an earlier commit); I left it in as it could conceivably be used at some point in the future for authenticating via the GH app (see https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps).  However if there's doubt about whether it's OK to include this, I'd prefer to just remove the script entirely.

Comment (A-Hilaly) [reply cd/lib/jwt.sh:4]: I'm not super sure with this either. I don't want to block you here, and i'm not again keeping this file here to use it in the future. I just know that to use third party software we need to publish an `ATTRIBUTIONS.md` file containing their respective licenses. e.g https://github.com/aws-controllers-k8s/runtime/blob/main/ATTRIBUTION.md

Comment (jljaco) [reply cd/lib/jwt.sh:4]: Removed the script, can just figure out the license question later if we end up wanting to use JWT

**Review by A-Hilaly — APPROVED** [review]
Merci beaucoup, Mr John!
/lgtm

**Pull request merged**
